modules/test/playwright/tests/frontend-data-set-web/main/advanced_drop_files.spec.ts

```diff
@@ -0,0 +1,215 @@
+			locator: Locator,
+			expectedClass?: string[]
+		) => {
+			await locator.dispatchEvent('dragstart', {dataTransfer});
+			await locator.dispatchEvent('dragenter', {dataTransfer});
+			await locator.dispatchEvent('dragover', {dataTransfer});
+
+			expectedClass &&
+				expect(
+					await containsClass(locator, expectedClass)
+				).toBeTruthy();
+		};
+
+		const dropFileOverLocator = async (
+			locator: Locator,
+			expectedText: string
+		) => {
+			await locator.dispatchEvent('dragstart', {dataTransfer});
+			await locator.dispatchEvent('dragenter', {dataTransfer});
+			await locator.dispatchEvent('dragover', {dataTransfer});
+			await locator.dispatchEvent('drop', {dataTransfer});
+			await locator.dispatchEvent('dragend', {dataTransfer});
+
+			await expect(fdsSamplePage.fileDropModal).toBeInViewport();
+			await expect(fdsSamplePage.fileDropModal).toContainText(
+				expectedText
+			);
+
+			await fdsSamplePage.fileDropModal.focus();
+
+			await page.keyboard.press('Escape');
+		};
+
+		const stopDragFileOverLocator = async (locator: Locator) => {
+			await locator.dispatchEvent('dragleave', {dataTransfer});
+			await locator.dispatchEvent('dragend', {dataTransfer});
+
+			await expect(locator).not.toHaveClass('drop-target');
+		};
+
+		await test.step('drag files over FDS main area highlights it', async () => {
+			await dragFileOverLocator(fdsSamplePage.fdsWrapper, [
+				'data-set-wrapper',
+				'visualization-mode-table',
+				'drop-target',
+			]);
+			await stopDragFileOverLocator(fdsSamplePage.fdsWrapper);
+		});
+
+		await test.step('drop files over FDS main area', async () => {
+			await dropFileOverLocator(
+				fdsSamplePage.fdsWrapper,
+				'no specific drop target'
+			);
+		});
+
+		const visualizationModes: Record<
+			VisualizationMode,
+			{
+				initialClass: string[];
+				itemLocatorContainer: Locator;
+				wrapperClass: string;
+			}
+		> = {
+			Cards: {
+				initialClass: ['card'],
+				itemLocatorContainer: fdsSamplePage.cards.items,
+				wrapperClass: 'visualization-mode-cards',
+			},
+			List: {
+				initialClass: ['list-group-item', 'list-group-item-flex'],
+				itemLocatorContainer: fdsSamplePage.list.items,
+				wrapperClass: 'visualization-mode-list',
+			},
+			Table: {
+				initialClass: [],
+				itemLocatorContainer: fdsSamplePage.table.bodyRows,
+				wrapperClass: 'visualization-mode-table',
+			},
+		};
+
+		for (const [
+			visualizationMode,
+			{initialClass, itemLocatorContainer, wrapperClass},
+		] of Object.entries(visualizationModes)) {
+			await fdsSamplePage.changeVisualizationMode(
+				visualizationMode as VisualizationMode
+			);
+
+			const blueItem = itemLocatorContainer
+				.filter({
+					hasText: 'sample 100.',
+				})
+				.first();
+
+			await test.step(`drag files over a droppable ${visualizationMode} item highlights it`, async () => {
+				await dragFileOverLocator(blueItem, [
+					'drop-target',
+					...initialClass,
+				]);
+
+				// FDS main area not highlighted
```

Comment:
Instead of inline comments, more Playwright-y is to have one more `test.step`, describing inner code.

Reply:
just a better test step description is fine